feat!: build/Actions improvements; Markup -> HTML (#14)
fsprojects/fsharp-cheatsheet: docs/fsharp-cheatsheet.md

> This cheatsheet glances over some of the common syntax of [F# 6.0](http://research.microsoft.com/en-us/um/cambridge/projects/fsharp/manual/spec.html). A link directly to [F# Documentation](https://learn.microsoft.com/en-us/dotnet/fsharp/). If you have any comments, corrections, or suggested additions, please open an issue or send a pull request to [https://github.com/fsprojects/fsharp-cheatsheet](https://github.com/fsprojects/fsharp-cheatsheet).

Reviewer: ```suggestion
This cheatsheet aims to succinctly cover the most important aspects of [F# 6.0](http://research.microsoft.com/en-us/um/cambridge/projects/fsharp/manual/spec.html).

The Microsoft [F# Documentation](https://learn.microsoft.com/en-us/dotnet/fsharp/) is complete and authoritative and has received a lot of love in recent years; it's well worth the time investment to read. But, only after you've got the lowdown here...

 If you have any comments, corrections, or suggested additions, please open an issue or send a pull request to [https://github.com/fsprojects/fsharp-cheatsheet](https://github.com/fsprojects/fsharp-cheatsheet). Questions are best addressed via the [F# slack](https://fsharp.org/guides/slack) or the [F# discord](https://discord.me/fsharp).
```

		| Node(l, _, r) -> 1 + max (depth l) (depth r)
		| Leaf -> 0

> F# Core has a few built-in discriminated unions for error handling, e.g., [Options](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/options) and [Results](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/results).

Reviewer: ```suggestion
F# Core has built-in discriminated unions for error handling, e.g., [`option`](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/options) and [`Result`](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/results).
```

<a name="ActivePatterns"></a>Active Patterns
---------------
*Single-case active patterns*:

    // Basic
    let (|EmailDomain|) (email: string) =
        let parts = email.Split '@'

>         parts[1]

Reviewer: This can be addressed later but we should make this real by handling cases where there's no `@`
There's a .NET class encapsulating email addresses too,  but using it in a cheatsheet might be a bit much

    let (EmailDomain emailDomain) = "yennefer@aretuza.org"  // emailDomain = 'aretuza.org'

    // As Parameters
    let (|Real|) (x: System.Numerics.Complex) =
        (x.Real, x.Imaginary)

>     let addReal (Real aa) (Real bb) =  // conversion done in the parameters

Reviewer: ```suggestion
    let addReal (Real (xr, _xi)) (Real (yr, _yi)) =  // conversion done in the parameters
        xr + yr
```
(depends if those concepts have been traversed at this point in the exercise)

Author: Thanks. I'll think on this one. It's two concepts; but, I agree, so much power comes from joining concepts that aren't obvious.

Reviewer: In general I avoid fst and snd and always deconstruct explicitly like this, ignoring the one I don't want.
The main exception to that is that I'll use `fst` for e.g. `xs |> Seq.map fst`
The main trigger for me suggesting anyting is that aa and bb dont mean anything, so I tend to call two equivalent args (esp where there is not real order to them) x and y
Main reasoning is that pattern matching is something you want to encourage use of as it nudges in good directions for code in general, whereas a sea of fst this snd that rarely ends up producing scannable code.

	    | Even -> printfn "%d is even" i
	    | Odd -> printfn "%d is odd" i

    let (|Phone|Email|) (s:string) =

>         if s.Contains '@' then Email $"Email: {s}" else Phone $"Phone: {s}"

Reviewer: ```suggestion
        if s.Contains '@' then Email s
        else Phone s
```

Author: I prefer the example as e.g. `Email $"Email: {s}"` to demonstrate that the argument to the identifier didn't need to match the parameter to the active pattern function.

Reviewer: My main thing was that for two cases doing equal amounts of work to put the else on its own line
Ambivalent about putting the prefix redundantly in the DU body (normally I'd expect a print function to pattern match and layer that on)
But I think the example in general is worth having so all good

> 

Reviewer: ```suggestion
Reference a .NET assembly:
```

	#r "../lib/FSharp.Markdown.dll"

Reference a nuget package

> 	#r "nuget: FSharp.Data, 6.3.0"

Reviewer: ```suggestion
    #r "nuget:Serilog.Sinks.Console" // latest production release
    #r "nuget: FSharp.Data, 6.3.0" // specific version
    #r "nuget:Equinox, *-*" // latest version, including `-alpha`, `-rc` version etc
```

> 

Reviewer: Reference a .NET assembly (`/` symbol is recommended for Mono compatibility).
```suggestion
Reference an assembly (very latest, including pre-release versions)

    #r "nuget:Equinox, *-*"
```
